src/state/service.rs

```diff
@@ -461,56 +451,65 @@ impl ServiceStore {
             }
         }
     }
 
     /// Adds the given service.
-    pub fn insert(&mut self, mut service: Service) {
-        // First add any staged service endpoints. Due to ordering issues, we may have received
-        // the workloads before their associated services.
+    pub fn insert(&mut self, service: Service) {
+        self.insert_internal(service, false)
+    }
+
+    /// insert_endpoint_update is like insert, but optimized for the case where we know only endpoints change.
+    pub fn insert_endpoint_update(&mut self, service: Service) {
+        self.insert_internal(service, true)
+    }
+
+    fn insert_internal(&mut self, mut service: Service, endpoint_update_only: bool) {
         let namespaced_hostname = service.namespaced_hostname();
-        if let Some(endpoints) = self.staged_services.remove(&namespaced_hostname) {
-            for (wip, ep) in endpoints {
-                if service.should_include_endpoint(ep.status) {
-                    service.endpoints.insert(wip.clone(), ep);
+        // If we're replacing an existing service, remove the old one from all data structures.
+        if !endpoint_update_only {
+            // First add any staged service endpoints. Due to ordering issues, we may have received
+            // the workloads before their associated services.
+            if let Some(endpoints) = self.staged_services.remove(&namespaced_hostname) {
+                for (wip, ep) in endpoints {
+                    if service.should_include_endpoint(ep.status) {
+                        service.endpoints.insert(wip.clone(), ep);
+                    }
                 }
             }
-        }
-
-        // If we're replacing an existing service, remove the old one from all data structures.
-        let _ = self.remove(&namespaced_hostname);
 
-        // Save values used for the indexes.
-        let vips = service.vips.clone();
-        let hostname = service.hostname.clone();
+            let _ = self.remove(&namespaced_hostname);
+        }
 
         // Create the Arc.
         let service = Arc::new(service);
+        let hostname = &service.hostname;
 
         // Map the vips to the service.
-        for vip in &vips {
+        for vip in &service.vips {
             self.by_vip.insert(vip.clone(), service.clone());
         }
 
         // Map the hostname to the service.
-        match self.by_host.get_mut(&hostname) {
+        match self.by_host.get_mut(hostname) {
             None => {
                 let _ = self.by_host.insert(hostname.clone(), vec![service.clone()]);
             }
             Some(services) => {
-                services.push(service.clone());
+                if let Some((cur, _)) = services
+                    .iter()
+                    .find_position(|s| s.namespace == service.namespace)
+                {
+                    // Service already exists; replace the slot
+                    services[cur] = service.clone()
```

Comment:
Does this implicitly drop 1 strong to the old Arc<Service> that was held by the vec? I suppose it must or it would leak

Reply:
yeah it will Drop the old one.